Added assertJsonStructure.

Zaboday opened the pull request:
Added assertJsonStructure tests.
CHANGELOG.md minor version up.

| Q             | A
| ------------- | ---
| Bug fix?      |  No
| New feature?  | Yes

## Description

Реализован assert подобный assertArrayStructure, который проверяет соответствие некой Json-строки определённой структуре - assertJsonStructure. assertJsonStructure получает на вход вместо массива JSON-строку которую декодирует и посылает в тот же assertArrayStructure.

## Checklist

- [ ] Вынес одинаковые тестовые данные для assertArrayStructure и assertJsonStructure в переменные.
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I wrote unit tests for my code
- [ ] I have made changes in [CHANGELOG.md](https://github.com/avto-dev/dev-tools/blob/master/CHANGELOG.md) file

tarampampam commented:
Previous tests should be not modificated.

Reallife commented on src/Tests/PHPUnit/Traits/AdditionalAssertionsTrait.php:
```diff
@@ -322,4 +322,22 @@ public static function assertArrayStructure($structure, $testing_array)
+     *
+     * @throws \InvalidArgumentException
+     */
+    public static function assertJsonStructure($structure, $testing_array_json)
```
Предлагаю второй параметр назвать просто `$testing_json`.

Zaboday replied:
ok

tarampampam replied:
`$json_string`

Reallife commented on tests/Tests/PHPUnit/Traits/AdditionalAssertionsTraitTest.php:
```diff
@@ -77,52 +77,69 @@ public function __toString()
-            'assertArrayStructure',
-            // Valid structures
+        /* @see AdditionalAssertionsTrait::assertArrayStructure */
+        $validStructures   = [
```
Примеры структур лучше вынести в отдельный метод.

tarampampam replied:
Also, in tests we should pass json-string as-is, without converting into arrays

Zaboday commented:
Fixes after Code review:
- Renamed assertJsonStructure parameter to **$json_string**
- Moved structures test arrays to structuresData() for reusage in structure methods.
- Moved testing Exeption "Invalid JSON given" to own test method to isolate expecting ecxeption

Reallife commented on tests/Tests/PHPUnit/Traits/AdditionalAssertionsTraitTest.php:
```diff
@@ -77,8 +77,49 @@ public function __toString()
+    }
+
+    /**
+     * @expectedException InvalidArgumentException
```
Давай будем явно указывать, что мы ждём `InvalidArgumentException` с сообщением `Invalid JSON given`. Используя методы `expectException` и `expectExceptionMessage` а не в аннотации. 
И предлагаю изменить сообщение на `Passed string has not valid JSON format`.

Reallife replied:
done

Reallife commented on tests/Tests/PHPUnit/Traits/AdditionalAssertionsTraitTest.php:
```diff
@@ -77,8 +77,49 @@ public function __toString()
+     *
+     * @return array
+     */
+    private function structuresData()
```
Метод лучше назвать `getStructuresData` - так точнее можно понять, что именно делает метод.

tarampampam commented on src/Tests/PHPUnit/Traits/AdditionalAssertionsTrait.php:
```diff
@@ -322,4 +322,22 @@ public static function assertArrayStructure($structure, $testing_array)
+    {
+        static::assertIsString($json_string);
+        $testing_array = \json_decode($json_string, true);
+        if (\json_last_error()) {
```
```php
if ((\json_last_error() !== JSON_ERROR_NONE || ! \is_array($testing_array)) { // ...
```

tarampampam commented on src/Tests/PHPUnit/Traits/AdditionalAssertionsTrait.php:
```diff
@@ -322,4 +322,22 @@ public static function assertArrayStructure($structure, $testing_array)
+     *
+     * @throws \InvalidArgumentException
+     */
+    public static function assertJsonStructure($structure, $json_string)
```
```php
($structure, $json_string, int $encode_options = 0)
```

tarampampam commented on src/Tests/PHPUnit/Traits/AdditionalAssertionsTrait.php:
```diff
@@ -322,4 +322,22 @@ public static function assertArrayStructure($structure, $testing_array)
+    public static function assertJsonStructure($structure, $json_string)
+    {
+        static::assertIsString($json_string);
+        $testing_array = \json_decode($json_string, true);
```
```php
$testing_array = \json_decode($json_string, true, $encode_options);
```

tarampampam commented on src/Tests/PHPUnit/Traits/AdditionalAssertionsTrait.php:
```diff
@@ -322,4 +322,22 @@ public static function assertArrayStructure($structure, $testing_array)
+     */
+    public static function assertJsonStructure($structure, $json_string)
+    {
+        static::assertIsString($json_string);
```
After this line add one empty line, please

tarampampam commented on src/Tests/PHPUnit/Traits/AdditionalAssertionsTrait.php:
```diff
@@ -322,4 +322,22 @@ public static function assertArrayStructure($structure, $testing_array)
+        if (\json_last_error()) {
+            throw new \InvalidArgumentException('Passed string has not valid JSON format.');
+        }
+        static::assertArrayStructure($structure, $testing_array);
```
Above this line add one empty line too, please

tarampampam commented on src/Tests/PHPUnit/Traits/AdditionalAssertionsTrait.php:
```diff
@@ -322,4 +322,22 @@ public static function assertArrayStructure($structure, $testing_array)
+     * @param array  $structure
+     * @param string $json_string
+     *
+     * @throws \InvalidArgumentException
```
```php
     * @throws AssertionFailedError
     * @throws InvalidArgumentException
```
